homework_solution.md

```diff
@@ -0,0 +1,474 @@
+# Part 1: Missing Station Data
+
+1. _**Write a query that fills out this table. Try your best to pick the correct station name for each ID. You may have to make some manual choices or editing based on the inconsistencies we've found. Do try to pick the correct name for each station ID based on how popular it is in the trip data.**_
+
+   ### Steps
+
+   1. Create an index on `from_station_id` and `to_station_id` on `trips` table to help query on those
+
+      ```
+      CREATE INDEX idx_from_station ON trips USING btree (from_station_id);
+      ```
+
+      ```
+      CREATE INDEX idx_to_station ON trips USING btree (to_station_id);
+      ```
+
+   2. Delete the exist `stations` table and recreate it with serial ids, so we can rely on auto-increment. (We could try and alter the table, but that seems to be an involved process, and our table being currently empty, it's easier to just recreate it)
+
+      ```
+      DROP TABLE stations;
+      ```
+
+      ```
+      CREATE TABLE stations
+      (
+         id SERIAL,
+         name VARCHAR(100)
+      );
+      ```
+
+      The effect of this can be perceived by running `SELECT nextval(pg_get_serial_sequence('stations', 'id'));`.
+      On the old table, that sequence, though present, wouldn't return a value. Now it does.
+
+   3. Get the first name for stations, assuming that as the correct one where there's more than one, combining the `from` and `to` column results
+
+      ```
+      SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
```

Comment:
[question]
What should you consider when using DISTINCT in an SQL query?  How can you mitigate those risks?

Reply:
Not sure I understand the question, so will try to share what I do know, and hope you'll let me know if it covers what you meant. 😄 

`DISTINCT` allows me to remove any duplicates from my query result. Duplicates in this context, meaning any records where the values I'm referring to in the query are the same, even if other columns excluded from the query would make those records different.
In this particular case, the query would serve as basis for an `INSERT` later on. Bringing multiple records, would mean the insert operation would fail when we tried to insert multiple records with the same `from_station_id`, as that's our primary key, and therefore expects unique values.

Reply:
That's correct, insofar as that's what DISTINCT does, but the risk I was looking for is that it has to compare all of the values to existing ones in order to determine if they're distinct or not.  One mitigation strategy is to use an index on the column, which you are already doing. :)

Reply:
Thanks for the extra info!!